In [77]:
import pandas as pd
import numpy as np
import datetime as dt
import plotly.express as px
import plotly.graph_objects as go

In [78]:
retail_data=pd.read_csv("data/online_retail_II.csv")
retail_data.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

           InvoiceDate  Price  Customer ID         Country  
0  2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1  2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2  2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3  2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4  2009-12-01 07:45:00   1.25      13085.0  United Kingdom

In [79]:
retail_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  object 
 1   StockCode    1067371 non-null  object 
 2   Description  1062989 non-null  object 
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  object 
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 65.1+ MB


1. Invoicedate column contains datetime info, while having type object, we need to separate column  into InvoiceDate_datetime

2. 1062689 rows but in customer id we have only 824364 so data is missing there, drop the rows with missing customer ids

3. customer id column type change from float to int

4. create a column name as total where price * quantity

In [80]:
retail_data["InvoiceDate_DT"]=pd.to_datetime(retail_data["InvoiceDate"])

retail_data=retail_data[retail_data["Customer ID"].notna()]
retail_data["Customer ID"]=retail_data['Customer ID'].astype(np.int64)
retail_data=retail_data.sort_values("InvoiceDate_DT")
retail_data["Total"]=retail_data["Quantity"]*retail_data["Price"]

In [81]:
retail_data.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

           InvoiceDate  Price  Customer ID         Country  \
0  2009-12-01 07:45:00   6.95        13085  United Kingdom   
1  2009-12-01 07:45:00   6.75        13085  United Kingdom   
2  2009-12-01 07:45:00   6.75        13085  United Kingdom   
3  2009-12-01 07:45:00   2.10        13085  United Kingdom   
4  2009-12-01 07:45:00   1.25        13085  United Kingdom   

       InvoiceDate_DT  Total  
0 2009-12-01 07:45:00   83.4  
1 2009-12-01 07:45:00   81.0  
2 2009-12-01 07:45:00   81.0  
3 2009-12-01 07:45:00  100.8  
4 2009-12-01 07:45:00   30.0

In [82]:
retail_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 824364 entries, 0 to 1067370
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   Invoice         824364 non-null  object        
 1   StockCode       824364 non-null  object        
 2   Description     824364 non-null  object        
 3   Quantity        824364 non-null  int64         
 4   InvoiceDate     824364 non-null  object        
 5   Price           824364 non-null  float64       
 6   Customer ID     824364 non-null  int64         
 7   Country         824364 non-null  object        
 8   InvoiceDate_DT  824364 non-null  datetime64[ns]
 9   Total           824364 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(2), object(5)
memory usage: 69.2+ MB


Analysis 


In [83]:
#Most and least expensive product 

most_expensive=retail_data.loc[retail_data["Price"]==retail_data["Price"].max()]
most_expensive

Invoice StockCode Description  Quantity          InvoiceDate    Price  \
748142  C556445         M      Manual        -1  2011-06-10 15:31:00  38970.0   

        Customer ID         Country      InvoiceDate_DT    Total  
748142        15098  United Kingdom 2011-06-10 15:31:00 -38970.0

In [84]:
least_expensive=retail_data.loc[retail_data["Price"]==retail_data["Price"].min()]
least_expensive

Invoice StockCode                     Description  Quantity  \
4674     489825     22076              6 RIBBONS EMPIRE          12   
6781     489998     48185             DOOR MAT FAIRY CAKE         2   
16107    490727         M                          Manual         1   
18739    490961     22142    CHRISTMAS CRAFT WHITE FAIRY         12   
18738    490961     22065  CHRISTMAS PUDDING TRINKET POT          1   
...         ...       ...                             ...       ...   
1004540  577129     22464     HANGING METAL HEART LANTERN         4   
1005014  577168         M                          Manual         1   
1006110  577314     23407  SET OF 2 TRAYS HOME SWEET HOME         2   
1011446  577696         M                          Manual         1   
1027583  578841     84826  ASSTD DESIGN 3D PAPER STICKERS     12540   

                 InvoiceDate  Price  Customer ID         Country  \
4674     2009-12-02 13:34:00    0.0        16126  United Kingdom   
6781     2009-12-03 11:19:00    0.0        15658  United Kingdom   
16107    2009-12-07 16:38:00    0.0        17231  United Kingdom   
18739    2009-12-08 15:25:00    0.0        14108  United Kingdom   
18738    2009-12-08 15:25:00    0.0        14108  United Kingdom   
...                      ...    ...          ...             ...   
1004540  2011-11-17 19:52:00    0.0        15602  United Kingdom   
1005014  2011-11-18 10:42:00    0.0        12603         Germany   
1006110  2011-11-18 13:23:00    0.0        12444          Norway   
1011446  2011-11-21 11:57:00    0.0        16406  United Kingdom   
1027583  2011-11-25 15:57:00    0.0        13256  United Kingdom   

             InvoiceDate_DT  Total  
4674    2009-12-02 13:34:00    0.0  
6781    2009-12-03 11:19:00    0.0  
16107   2009-12-07 16:38:00    0.0  
18739   2009-12-08 15:25:00    0.0  
18738   2009-12-08 15:25:00    0.0  
...                     ...    ...  
1004540 2011-11-17 19:52:00    0.0  
1005014 2011-11-18 10:42:00    0.0  
1006110 2011-11-18 13:23:00    0.0  
1011446 2011-11-21 11:57:00    0.0  
1027583 2011-11-25 15:57:00    0.0  

[71 rows x 10 columns]

In [85]:
len(least_expensive['Description'].unique())

61

In [86]:
least_expensive[least_expensive['Description']=="36 FOIL STAR CAKE CASES "]

Invoice StockCode               Description  Quantity  \
840208  564651     22955  36 FOIL STAR CAKE CASES        144   

                InvoiceDate  Price  Customer ID      Country  \
840208  2011-08-26 14:19:00    0.0        14646  Netherlands   

            InvoiceDate_DT  Total  
840208 2011-08-26 14:19:00    0.0

In [87]:
least_expensive_group =least_expensive.groupby(["Description","Price"],as_index=False)["Quantity"].agg("sum")
least_expensive_group

Description  Price  Quantity
0                    FLAMINGO LIGHTS    0.0        24
1         OVAL WALL MIRROR DIAMANTE     0.0         1
2           36 FOIL STAR CAKE CASES     0.0       144
3                 6 RIBBONS EMPIRE      0.0        12
4       ADVENT CALENDAR GINGHAM SACK    0.0         4
..                               ...    ...       ...
56  SET/5 RED SPOTTY LID GLASS BOWLS    0.0         2
57         TV DINNER TRAY DOLLY GIRL    0.0         9
58           This is a test product.    0.0        10
59  VINTAGE CREAM CAT FOOD CONTAINER    0.0         1
60        VINTAGE GLASS COFFEE CADDY    0.0         1

[61 rows x 3 columns]

In [88]:
least_expensive_group["Description"].unique()

array([' FLAMINGO LIGHTS', ' OVAL WALL MIRROR DIAMANTE ',
       '36 FOIL STAR CAKE CASES ', '6 RIBBONS EMPIRE  ',
       'ADVENT CALENDAR GINGHAM SACK',
       'AIRLINE BAG VINTAGE JET SET WHITE',
       'ANTIQUE GLASS HEART DECORATION ', 'ANTIQUE LILY FAIRY LIGHTS',
       'ASSTD DESIGN 3D PAPER STICKERS', 'BISCUIT TIN VINTAGE CHRISTMAS',
       'BREAD BIN DINER STYLE RED ', 'CAKE STAND LACE WHITE',
       'CAST IRON HOOK GARDEN FORK', 'CAST IRON HOOK GARDEN TROWEL',
       'CERAMIC BOWL WITH LOVE HEART DESIGN',
       'CHARLOTTE BAG , SUKI DESIGN', "CHILDREN'S APRON DOLLY GIRL ",
       'CHILDS BREAKFAST SET CIRCUS PARADE',
       'CHRISTMAS CRAFT WHITE FAIRY ', 'CHRISTMAS PUDDING TRINKET POT ',
       'DOLLY GIRL LUNCH BOX', 'DOOR MAT FAIRY CAKE',
       'DOORMAT HOME SWEET HOME BLUE ', 'FAIRY CAKES NOTEBOOK A6 SIZE',
       'GLASS CLOCHE SMALL', 'HANGING METAL BIRD BATH',
       'HANGING METAL HEART LANTERN', 'HEART GARLAND RUSTIC PADDED',
       'IVORY KITCHEN SCALES', 'JAM MAKIN

The most expensive product in the dataset has weird description, "MANUAL", is prices at 38970.0., quantity is -1, looks like some one returned an item priced that high or did not pay for it yet. 

finding the least expensive product needs one extra step, because we have 61 unique description of products priced at 0.0pounds. discard the zero-priced product 
 

In [89]:
retail_non_zero_price=retail_data.loc[retail_data["Price"]!=0]

least_expensive_not_null=retail_non_zero_price.loc[retail_non_zero_price["Price"]==retail_non_zero_price["Price"].min()]

least_expensive_not_null_group= least_expensive_not_null.groupby(["Description","Price"],as_index=False)["Quantity"].agg("sum")

In [90]:
least_expensive_not_null

Invoice     StockCode                 Description  Quantity  \
62299   494914          PADS  PADS TO MATCH ALL CUSHIONS         1   
74731   496222          PADS  PADS TO MATCH ALL CUSHIONS         1   
77702   496473          PADS  PADS TO MATCH ALL CUSHIONS         1   
79794   496643          PADS  PADS TO MATCH ALL CUSHIONS         1   
90798   497935          PADS  PADS TO MATCH ALL CUSHIONS         1   
97716   498562          PADS  PADS TO MATCH ALL CUSHIONS         1   
101718  499056          PADS  PADS TO MATCH ALL CUSHIONS         1   
104480  499399          PADS  PADS TO MATCH ALL CUSHIONS         1   
123947  501176          PADS  PADS TO MATCH ALL CUSHIONS         1   
156809  504332          PADS  PADS TO MATCH ALL CUSHIONS         1   
178629  506303          PADS  PADS TO MATCH ALL CUSHIONS         1   
341826  522766          PADS  PADS TO MATCH ALL CUSHIONS         1   
440930  531433          PADS  PADS TO MATCH ALL CUSHIONS         1   
468280  533528          PADS  PADS TO MATCH ALL CUSHIONS         1   
682656  550193          PADS  PADS TO MATCH ALL CUSHIONS         1   
804506  561226          PADS  PADS TO MATCH ALL CUSHIONS         1   
885332  568200          PADS  PADS TO MATCH ALL CUSHIONS         1   
887202  568375  BANK CHARGES                Bank Charges         1   

                InvoiceDate  Price  Customer ID         Country  \
62299   2010-01-19 17:04:00  0.001        16705  United Kingdom   
74731   2010-01-29 13:53:00  0.001        13583  United Kingdom   
77702   2010-02-01 15:38:00  0.001        17350  United Kingdom   
79794   2010-02-03 11:58:00  0.001        13408  United Kingdom   
90798   2010-02-15 10:47:00  0.001        13408  United Kingdom   
97716   2010-02-21 12:03:00  0.001        15182  United Kingdom   
101718  2010-02-24 13:46:00  0.001        13765  United Kingdom   
104480  2010-02-26 13:26:00  0.001        14459  United Kingdom   
123947  2010-03-15 11:00:00  0.001        14857  United Kingdom   
156809  2010-04-12 16:30:00  0.001        12671         Germany   
178629  2010-04-29 10:43:00  0.001        14249  United Kingdom   
341826  2010-09-16 13:58:00  0.001        15182  United Kingdom   
440930  2010-11-08 11:54:00  0.001        13317  United Kingdom   
468280  2010-11-17 15:08:00  0.001        15750  United Kingdom   
682656  2011-04-15 09:27:00  0.001        13952  United Kingdom   
804506  2011-07-26 10:13:00  0.001        15618  United Kingdom   
885332  2011-09-25 14:58:00  0.001        16198  United Kingdom   
887202  2011-09-26 17:01:00  0.001        13405  United Kingdom   

            InvoiceDate_DT  Total  
62299  2010-01-19 17:04:00  0.001  
74731  2010-01-29 13:53:00  0.001  
77702  2010-02-01 15:38:00  0.001  
79794  2010-02-03 11:58:00  0.001  
90798  2010-02-15 10:47:00  0.001  
97716  2010-02-21 12:03:00  0.001  
101718 2010-02-24 13:46:00  0.001  
104480 2010-02-26 13:26:00  0.001  
123947 2010-03-15 11:00:00  0.001  
156809 2010-04-12 16:30:00  0.001  
178629 2010-04-29 10:43:00  0.001  
341826 2010-09-16 13:58:00  0.001  
440930 2010-11-08 11:54:00  0.001  
468280 2010-11-17 15:08:00  0.001  
682656 2011-04-15 09:27:00  0.001  
804506 2011-07-26 10:13:00  0.001  
885332 2011-09-25 14:58:00  0.001  
887202 2011-09-26 17:01:00  0.001

In [91]:
least_expensive_not_null_group

Description  Price  Quantity
0                Bank Charges  0.001         1
1  PADS TO MATCH ALL CUSHIONS  0.001        17

so the cheapest product is "PADS TO MATCH ALL CUSHIONS" and "BANK CHARGES" price 0.0001 pounds

Customer Analysis: Total Purchase Amount

Next, we will find out 10 customer who spent most in the online shopping for two years.Those are the most valuable clients, Also I'll define customer spending least

In [92]:
retail_customer=retail_data.groupby(["Customer ID","Country"],as_index=False)["Total"].agg("sum")
retail_customer.head()


Customer ID         Country    Total
0        12346  United Kingdom   -64.68
1        12347         Iceland  5633.32
2        12348         Finland  2019.40
3        12349           Italy  4404.54
4        12350          Norway   334.40

In [93]:
retail_customer_asc=retail_customer.sort_values("Total").head(10)
retail_customer_asc

Customer ID         Country     Total
5066        17399  United Kingdom -25111.09
585         12918  United Kingdom -10953.50
3516        15849  United Kingdom  -5876.34
3427        15760          Norway  -5795.87
4648        16981  United Kingdom  -4620.86
3818        16151  United Kingdom  -4217.59
1730        14063  United Kingdom  -3767.20
5690        18023  United Kingdom  -3248.86
4680        17013  United Kingdom  -3224.76
2869        15202  United Kingdom  -2570.18

In [94]:
retail_customer_desc=retail_customer.sort_values("Total",ascending=False).head(10)
retail_customer_desc

Customer ID         Country      Total
5769        18102  United Kingdom  598215.22
2313        14646     Netherlands  523342.07
1823        14156            EIRE  296564.69
2578        14911            EIRE  270248.53
5117        17450  United Kingdom  233579.39
1361        13694  United Kingdom  190825.52
5178        17511  United Kingdom  171885.98
72          12415       Australia  143269.29
4351        16684  United Kingdom  141502.25
2728        15061  United Kingdom  136391.48

In [95]:
pip install nbformat --upgrade


[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [96]:
fig=go.Figure(data=[
    go.Bar(name="Customer with max total purchase amount",
           x=retail_customer_desc["Customer ID"].astype(str),
           y=retail_customer_desc["Total"],
           opacity=1,
           hovertext=retail_customer_desc['Country'],
           marker={'color':retail_customer_desc["Total"],
                   "colorscale":'Rainbow'})
])

fig.update_traces(texttemplate='£%{y:.3s}', textposition='inside')

fig.update_layout(title='Customers With Max Total Purchase Amount',
                  title_x=0.45,
                  xaxis_title="Customer ID",
                  yaxis_title="Total Amount, £",
                  plot_bgcolor='white')
fig.show()

In [97]:
fig = go.Figure(data=[
    go.Bar(name='Customers With Min Total Purchase Amount', 
           x=retail_customer_asc['Customer ID'].astype(str), 
           y=retail_customer_asc['Total'],
           marker_opacity=1,
           hovertext=retail_customer_desc['Country'],
           marker={'color': retail_customer_desc['Total'],
                   'colorscale': 'Rainbow'}
          )
    ])
fig.update_xaxes(side="top")

fig.update_traces(texttemplate='£%{y:.3s}', textposition='inside')

fig.update_layout(title='Customers With Min Total Purchase Amount',
                  title_x=0.45,
                  title_y=0.97,
                  xaxis_title="Customer ID",
                  yaxis_title="Total Amount Returned, £",
                  plot_bgcolor='white')
fig.show()

What have we learned from the charts?

Customer with ID 18102 has spent most money in our online store during two years. He comes from United Kingdom, and has paid in total £598k. Another customer from Netherlands comes close to that number - he has spent £523k. In general the top 10 consists of customers from UK, EIRE, Netherlands and Australia - all countries with mature economies.

On the chart showing customers who spent the least, we can see only negative numbers. From our dataset alone, it seems that those are returns. The largest total negative sum is £25k, and the customer with this returned amount comes from UK as well. It is curious that in general customers who have largest negative values here come from the same countries as customers who have paid most.

Countries Analysis: Total Purchase Amount
Here, I will find 10 countries, total sum of orders from which was the largest, and also 10 countries, where total purchases were minimal.

In [98]:
retail_data_country_purchase=retail_data.groupby(['Country'],as_index=False)['Total'].agg('sum')

retail_data_country_purchase_asc=retail_data_country_purchase.sort_values("Total").head(10)

retail_data_country_purchase_desc=retail_data_country_purchase.sort_values("Total",ascending=False).head(10)

In [99]:
fig = go.Figure(data=[
    go.Bar(name='Countries With Max Total Purchase Amount', 
           x=retail_data_country_purchase_desc['Country'].astype(str), 
           y=retail_data_country_purchase_desc['Total'],
           opacity=1,
           marker={'color': retail_data_country_purchase_desc['Total'],
                   'colorscale': 'Rainbow'})
    ])

fig.update_traces(texttemplate='£%{y:.3s}', textposition='outside')

fig.update_layout(title='Countries With Max Total Purchase Sum',
                  title_x=0.45,
                  xaxis_title="Countries",
                  yaxis_title="Total Amount, £",
                  plot_bgcolor='white')
fig.show()

In [100]:
fig = go.Figure(data=[
    go.Bar(name='Countries With Min Total Purchase Amount', 
           x=retail_data_country_purchase_asc['Country'].astype(str), 
           y=retail_data_country_purchase_asc['Total'],
           marker_opacity=1,
           marker={'color': retail_data_country_purchase_desc['Total'],
                   'colorscale': 'Rainbow'}
          )
    ])

fig.update_traces(texttemplate='£%{y:.3s}', textposition='inside')

fig.update_layout(title='Countries With Min Total Purchase Amount',
                  title_x=0.45,
                  title_y=0.97,
                  xaxis_title="Countries",
                  yaxis_title="Total Amount, £",
                  plot_bgcolor='white')
fig.show()

If we look at the 10 countries, where maximal purchase totals originate from, it is very clear that United Kingdom is a leader with purchase amount of £13.8M. The closest amount to that is £579k - that much was purchased by EIRE customers. It is 23 times less! One other interesting insight is that all countries in the top ten are located in Europe.

The country with the smallest purchase value is Saudi Arabia. Customers from Saudi Arabia have spent only £131 in the online shop in two years! While generally countries with smaller purchased amounts are situated outside of Europe, there is an interesting exception: Czech Republic. If I had more data, I would love to dig deeper into the reasons of this fact!

In [101]:
#unique customer Number 

retail_data_customer_unique=retail_data.groupby(["Country"],as_index=False)['Customer ID'].agg({'Customer ID':'nunique'})

retail_data_customer_unique.rename(columns={'Customer ID':"Customer_Count"},inplace=True)

retail_data_customer_unique_asc=retail_data_customer_unique.sort_values('Customer_Count').head(10)

retail_data_customer_unique_desc=retail_data_customer_unique.sort_values('Customer_Count',ascending=False).head(10)

/var/folders/j9/2xjjcspd5r3dz1m4ll_0c7m80000gn/T/ipykernel_9778/509398045.py:3: FutureWarning:

Passing a dictionary to SeriesGroupBy.agg is deprecated and will raise in a future version of pandas. Pass a list of aggregations instead.



In [102]:
fig = go.Figure(data=[
    go.Bar(name='Countries with Largest Number of Unique Customers', 
           x=retail_data_customer_unique_desc['Country'].astype(str), 
           y=retail_data_customer_unique_desc['Customer_Count'],
           marker_opacity=1,
           marker={'color': retail_data_customer_unique_desc['Customer_Count'],
                   'colorscale': 'Rainbow'})
    ])

fig.update_traces(texttemplate='%{y:.3s}', textposition='outside')

fig.update_layout(title='Countries with Largest Number of Unique Customers',
                  title_x=0.45,
                  xaxis_title="Countries",
                  yaxis_title="Number of Unique Customers",
                  plot_bgcolor='white')
fig.show()

In [103]:
fig = go.Figure(data=[
    go.Bar(name='Countries with Smallest Number of Unique Customers', 
           x=retail_data_customer_unique_asc['Country'].astype(str), 
           y=retail_data_customer_unique_asc['Customer_Count'],
           marker_opacity=1,
           marker={'color': retail_data_customer_unique_desc['Customer_Count'],
                   'colorscale': 'Rainbow'}
          )
    ])

fig.update_traces(texttemplate='%{y:.3s}', textposition='outside')

fig.update_layout(title='Countries with Smallest Number of Unique Customers',
                  title_x=0.45,
                  title_y=0.97,
                  xaxis_title="Countries",
                  yaxis_title="Number of Unique Customers",
                  plot_bgcolor='white')
fig.show()

In UK, 5.41k of unique customers have made purchases in two years, while Germany and France follow up with only 107 customers and 95 customers correspondingly. The difference between top position and the followers is indeed huge. Note that we have spotted the same while analysing purchases for each country.
Also there are many countries with a single unique customer, such as Thailand, Nigeria, Iceland, Saudi Arabia etc. Here, Czech Republic is again an outlier in the list of Asian, South American and African countries, but now we see that also Lithuania and Iceland have only one customer. While Iceland has a relatively small population, it would be interesting to investigate and compare the reasons for Czech Republic and Lithuania.

Products Analysis: Quantity
Next, I will find 10 bestselling products and 10 products which have top number of returns.

I will remove products with description 'Discount' and 'CRUK Commission' to concentrate on the representative data.

In [104]:
retail_data_products=retail_data.groupby(["Description"],as_index=False)["Quantity"].agg('sum')

retail_data_product=retail_data_products.loc[~retail_data_products['Description'].isin(['CRUK Commision','Discount'])]

retail_data_product_asc=retail_data_product.sort_values('Quantity').head(10)

retail_data_product_asc

Description  Quantity
5040                  WHITE CHERRY LIGHTS      -105
4391                 SILVER CHERRY LIGHTS       -96
4793  TREE OF NOAH FESTIVE SCENTED CANDLE       -34
1186                      CRUK Commission       -16
4093  SET 6 MINI SUSHI SET FRIDGE MAGNETS       -12
5087   WHITE SCANDINAVIAN HEART CHRISTMAS       -11
4882      VINTAGE BLUE VACUUM FLASK 0.5L        -10
473                   BLACK CHERRY LIGHTS        -8
1624                 FEATHER HEART LIGHTS        -8
5075           WHITE ORCHID FLOWER LIGHTS        -6

In [105]:
retail_data_product_asc['Quantity']=retail_data_product_asc['Quantity']*(-1)

retail_data_product_asc.head(10)

Description  Quantity
5040                  WHITE CHERRY LIGHTS       105
4391                 SILVER CHERRY LIGHTS        96
4793  TREE OF NOAH FESTIVE SCENTED CANDLE        34
1186                      CRUK Commission        16
4093  SET 6 MINI SUSHI SET FRIDGE MAGNETS        12
5087   WHITE SCANDINAVIAN HEART CHRISTMAS        11
4882      VINTAGE BLUE VACUUM FLASK 0.5L         10
473                   BLACK CHERRY LIGHTS         8
1624                 FEATHER HEART LIGHTS         8
5075           WHITE ORCHID FLOWER LIGHTS         6

In [106]:
retail_data_product_desc = retail_data_product.sort_values('Quantity', ascending=False).head(10)

In [107]:
fig = go.Figure(data=[
    go.Bar(name='10 Bestselling Products', 
           x=retail_data_product_desc['Description'].astype(str), 
           y=retail_data_product_desc['Quantity'],
           marker_opacity=1,
           marker={'color': retail_data_product_desc['Quantity'],
                   'colorscale': 'Rainbow'})
    ])

fig.update_traces(texttemplate='%{y:.3s}', textposition='outside')
fig.update_layout(barmode='group', showlegend=False)

fig.update_layout(title='10 Bestselling Products',
                  title_x=0.45,
                  xaxis_title="Products",
                  yaxis_title="Total Quantity Sold",
                  plot_bgcolor='white')
fig.show()

In [108]:
fig = go.Figure(data=[
    go.Bar(name='10 Products With Most Returns', 
           x=retail_data_product_asc['Description'].astype(str), 
           y=retail_data_product_asc['Quantity'],
           marker_opacity=1,
           marker={'color': retail_data_product_asc['Quantity'],
                   'colorscale': 'Rainbow'}
          )
    ])

fig.update_traces(texttemplate='%{y:.3s}', textposition='outside')
fig.update_layout(barmode='group', showlegend=False)

fig.update_layout(title='10 Products With Most Returns',
                  title_x=0.45,
                  title_y=0.97,
                  xaxis_title="Products",
                  yaxis_title="Total Quantity Returned",
                  plot_bgcolor='white')
fig.show()

We can see that World War II Gliders are clear bestsellers (the store has sold 107k of them in two years). However quite utilitary things also sell pretty well: mostly in the top ten different holders and cake cases prevail. Also a purse and a jumbo bag can be found there.

The product with most returns is White Cherry Lights (105 of them were returned in two years). It is intereting that many decor items and gifts have made it to this list: various candles, flasks, pots, magnets and lights are here.

**Cohort Analysis**

Cohort analysis is generally used to show behavious of users, or in our case, customers, over time. It is an incredibly useful instrument to find some insights about customer experiences, and to understand how to improve that.

I will use cohort analysis to investigate retention rate and average amount of products sold, and choose to use quarters of the year to define my cohorts. I will start by slightly transforming my data.

Our dataset covers a timeframe of two years (2010-2011) and additionally December 2009. As December 2009 does not give us a full picture about the quarter, I will remove it.

In order to make calculations easily, each quarter is mapped to an ID.

I also define my cohort quarter, and calculate cohort index as a difference (number of quarters) between each quarter and the first quarter (having ID 0).

In [109]:
retail_data.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

           InvoiceDate  Price  Customer ID         Country  \
0  2009-12-01 07:45:00   6.95        13085  United Kingdom   
1  2009-12-01 07:45:00   6.75        13085  United Kingdom   
2  2009-12-01 07:45:00   6.75        13085  United Kingdom   
3  2009-12-01 07:45:00   2.10        13085  United Kingdom   
4  2009-12-01 07:45:00   1.25        13085  United Kingdom   

       InvoiceDate_DT  Total  
0 2009-12-01 07:45:00   83.4  
1 2009-12-01 07:45:00   81.0  
2 2009-12-01 07:45:00   81.0  
3 2009-12-01 07:45:00  100.8  
4 2009-12-01 07:45:00   30.0

In [110]:
retail_data=retail_data[retail_data['InvoiceDate_DT'].dt.year!=2009]

retail_data['InvoiceQuater']=('Q' + retail_data['InvoiceDate_DT'].dt.quarter.astype(str) + '/' + retail_data['InvoiceDate_DT'].dt.year.astype(str))

In [111]:
retail_data.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  
45228 2010-01-04 09:24:00  22.50       Q1/2010  
45229 2010-01-04 09:43:00  -4.25       Q1/2010  
45230 2010-01-04 09:53:00  22.50       Q1/2010  
45238 2010-01-04 10:28:00  83.40       Q1/2010  
45237 2010-01-04 10:28:00   8.50       Q1/2010

In [112]:
retail_data["InvoiceQuater"].unique()

array(['Q1/2010', 'Q2/2010', 'Q3/2010', 'Q4/2010', 'Q1/2011', 'Q2/2011',
       'Q3/2011', 'Q4/2011'], dtype=object)

In [113]:
quarters_map=dict(zip(retail_data["InvoiceQuater"].unique(),range(len(retail_data['InvoiceQuater'].unique()))))

In [114]:
quarters_map

{'Q1/2010': 0,
 'Q2/2010': 1,
 'Q3/2010': 2,
 'Q4/2010': 3,
 'Q1/2011': 4,
 'Q2/2011': 5,
 'Q3/2011': 6,
 'Q4/2011': 7}

In [115]:
retail_data['InvoiceQuarterID']=retail_data['InvoiceQuater'].map(quarters_map)

retail_data.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  InvoiceQuarterID  
45228 2010-01-04 09:24:00  22.50       Q1/2010                 0  
45229 2010-01-04 09:43:00  -4.25       Q1/2010                 0  
45230 2010-01-04 09:53:00  22.50       Q1/2010                 0  
45238 2010-01-04 10:28:00  83.40       Q1/2010                 0  
45237 2010-01-04 10:28:00   8.50       Q1/2010                 0

In [116]:
retail_data["CohortQuarterID"]=retail_data.groupby('Customer ID')["InvoiceQuarterID"].transform('min')

retail_data.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  InvoiceQuarterID  \
45228 2010-01-04 09:24:00  22.50       Q1/2010                 0   
45229 2010-01-04 09:43:00  -4.25       Q1/2010                 0   
45230 2010-01-04 09:53:00  22.50       Q1/2010                 0   
45238 2010-01-04 10:28:00  83.40       Q1/2010                 0   
45237 2010-01-04 10:28:00   8.50       Q1/2010                 0   

       CohortQuarterID  
45228                0  
45229                0  
45230                0  
45238                0  
45237                0

In [117]:
retail_data.tail(30)

Invoice StockCode                          Description  Quantity  \
1067338  581585     22113          GREY HEART HOT WATER BOTTLE         4   
1067337  581585     84945   MULTI COLOUR SILVER T-LIGHT HOLDER        24   
1067336  581585     84879        ASSORTED COLOUR BIRD ORNAMENT        16   
1067335  581585     23084                   RABBIT NIGHT LIGHT        12   
1067334  581585     84832     ZINC WILLIE WINKIE  CANDLE STICK        24   
1067333  581585     22460       EMBOSSED GLASS TEALIGHT HOLDER        12   
1067332  581585     22178      VICTORIAN GLASS HANGING T-LIGHT        12   
1067331  581585     22915        ASSORTED BOTTLE TOP  MAGNETS         24   
1067339  581585     23356                LOVE HOT WATER BOTTLE         3   
1067330  581585     22481       BLACK TEA TOWEL CLASSIC DESIGN        12   
1067354  581586     20685                DOORMAT RED RETROSPOT        10   
1067353  581586     21217        RED RETROSPOT ROUND CAKE TINS        24   
1067351  581586     22061  LARGE CAKE STAND  HANGING STRAWBERY         8   
1067352  581586     23275     SET OF 3 HANGING OWLS OLLIE BEAK        24   
1067363  581587     22629                  SPACEBOY LUNCH BOX         12   
1067368  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
1067367  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
1067366  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
1067365  581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
1067364  581587     23256          CHILDRENS CUTLERY SPACEBOY          4   
1067362  581587     22367      CHILDRENS APRON SPACEBOY DESIGN         8   
1067355  581587     22631             CIRCUS PARADE LUNCH BOX         12   
1067360  581587     22726           ALARM CLOCK BAKELIKE GREEN         4   
1067359  581587     22727            ALARM CLOCK BAKELIKE RED          4   
1067358  581587     22728            ALARM CLOCK BAKELIKE PINK         4   
1067357  581587     22555            PLASTERS IN TIN STRONGMAN        12   
1067356  581587     22556       PLASTERS IN TIN CIRCUS PARADE         12   
1067369  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
1067361  581587     22730           ALARM CLOCK BAKELIKE IVORY         4   
1067370  581587      POST                              POSTAGE         1   

                 InvoiceDate  Price  Customer ID         Country  \
1067338  2011-12-09 12:31:00   4.25        15804  United Kingdom   
1067337  2011-12-09 12:31:00   0.85        15804  United Kingdom   
1067336  2011-12-09 12:31:00   1.69        15804  United Kingdom   
1067335  2011-12-09 12:31:00   2.08        15804  United Kingdom   
1067334  2011-12-09 12:31:00   0.85        15804  United Kingdom   
1067333  2011-12-09 12:31:00   1.25        15804  United Kingdom   
1067332  2011-12-09 12:31:00   1.95        15804  United Kingdom   
1067331  2011-12-09 12:31:00   0.19        15804  United Kingdom   
1067339  2011-12-09 12:31:00   5.95        15804  United Kingdom   
1067330  2011-12-09 12:31:00   0.39        15804  United Kingdom   
1067354  2011-12-09 12:49:00   7.08        13113  United Kingdom   
1067353  2011-12-09 12:49:00   8.95        13113  United Kingdom   
1067351  2011-12-09 12:49:00   2.95        13113  United Kingdom   
1067352  2011-12-09 12:49:00   1.25        13113  United Kingdom   
1067363  2011-12-09 12:50:00   1.95        12680          France   
1067368  2011-12-09 12:50:00   4.15        12680          France   
1067367  2011-12-09 12:50:00   4.15        12680          France   
1067366  2011-12-09 12:50:00   2.10        12680          France   
1067365  2011-12-09 12:50:00   0.85        12680          France   
1067364  2011-12-09 12:50:00   4.15        12680          France   
1067362  2011-12-09 12:50:00   1.95        12680          France   
1067355  2011-12-09 12:50:00   1.95        12680          France   
1067360  2011-12-09 12:50:00   3.75        12680          France   
1067359  2011-12-09

In [118]:
retail_data[retail_data["Customer ID"]==13113]

Invoice StockCode                          Description  Quantity  \
141805   502869     22131     FOOD CONTAINER SET 3 LOVE HEART          4   
141804   502869     21452                  TOADSTOOL MONEY BOX         1   
141803   502869     22073               RETRO SPOT STORAGE JAR         2   
141802   502869     22057      CERAMIC PLATE STRAWBERRY DESIGN         4   
141801   502869     22063  CERAMIC BOWL WITH STRAWBERRY DESIGN         2   
...         ...       ...                                  ...       ...   
1025344  578699     22457      NATURAL SLATE HEART CHALKBOARD         48   
1067354  581586     20685                DOORMAT RED RETROSPOT        10   
1067353  581586     21217        RED RETROSPOT ROUND CAKE TINS        24   
1067351  581586     22061  LARGE CAKE STAND  HANGING STRAWBERY         8   
1067352  581586     23275     SET OF 3 HANGING OWLS OLLIE BEAK        24   

                 InvoiceDate  Price  Customer ID         Country  \
141805   2010-03-28 16:13:00   1.95        13113  United Kingdom   
141804   2010-03-28 16:13:00   2.95        13113  United Kingdom   
141803   2010-03-28 16:13:00   3.75        13113  United Kingdom   
141802   2010-03-28 16:13:00   1.49        13113  United Kingdom   
141801   2010-03-28 16:13:00   2.95        13113  United Kingdom   
...                      ...    ...          ...             ...   
1025344  2011-11-25 10:22:00   2.55        13113  United Kingdom   
1067354  2011-12-09 12:49:00   7.08        13113  United Kingdom   
1067353  2011-12-09 12:49:00   8.95        13113  United Kingdom   
1067351  2011-12-09 12:49:00   2.95        13113  United Kingdom   
1067352  2011-12-09 12:49:00   1.25        13113  United Kingdom   

             InvoiceDate_DT   Total InvoiceQuater  InvoiceQuarterID  \
141805  2010-03-28 16:13:00    7.80       Q1/2010                 0   
141804  2010-03-28 16:13:00    2.95       Q1/2010                 0   
141803  2010-03-28 16:13:00    7.50       Q1/2010                 0   
141802  2010-03-28 16:13:00    5.96       Q1/2010                 0   
141801  2010-03-28 16:13:00    5.90       Q1/2010                 0   
...                     ...     ...           ...               ...   
1025344 2011-11-25 10:22:00  122.40       Q4/2011                 7   
1067354 2011-12-09 12:49:00   70.80       Q4/2011                 7   
1067353 2011-12-09 12:49:00  214.80       Q4/2011                 7   
1067351 2011-12-09 12:49:00   23.60       Q4/2011                 7   
1067352 2011-12-09 12:49:00   30.00       Q4/2011                 7   

         CohortQuarterID  
141805                 0  
141804                 0  
141803                 0  
141802                 0  
141801                 0  
...                  ...  
1025344                0  
1067354                0  
1067353                0  
1067351                0  
1067352                0  

[498 rows x 13 columns]

In [119]:
retail_data[retail_data["Customer ID"]==12680]


Invoice StockCode                          Description  Quantity  \
831253   563712     22029               SPACEBOY BIRTHDAY CARD        12   
831252   563712     22367      CHILDRENS APRON SPACEBOY DESIGN         8   
831254   563712     23256          CHILDRENS CUTLERY SPACEBOY          4   
831251   563712     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
831259   563712      POST                              POSTAGE         2   
831257   563712     22099            CARAVAN SQUARE TISSUE BOX        36   
831258   563712     22326  ROUND SNACK BOXES SET OF4 WOODLAND          6   
831255   563712     22728            ALARM CLOCK BAKELIKE PINK         4   
831250   563712     22899         CHILDREN'S APRON DOLLY GIRL          6   
831256   563712     22727            ALARM CLOCK BAKELIKE RED          4   
831248   563712     22712                     CARD DOLLY GIRL         12   
831247   563712     22979              PANTRY WASHING UP BRUSH        12   
831246   563712     22980               PANTRY SCRUBBING BRUSH        12   
831245   563712     22629                  SPACEBOY LUNCH BOX         12   
831244   563712     22555            PLASTERS IN TIN STRONGMAN        12   
831243   563712     23190  BUNDLE OF 3 SCHOOL EXERCISE BOOKS          12   
831242   563712     22037                  ROBOT BIRTHDAY CARD        12   
831241   563712     21986     PACK OF 12 PINK POLKADOT TISSUES        24   
831240   563712     21981         PACK OF 12 WOODLAND TISSUES         24   
831249   563712     22630                 DOLLY GIRL LUNCH BOX        12   
859960   566256      POST                              POSTAGE         1   
859959   566256     20751     FUNKY WASHING UP GLOVES ASSORTED        12   
859958   566256     21208          PASTEL COLOUR HONEYCOMB FAN        12   
859955   566256     22726           ALARM CLOCK BAKELIKE GREEN         4   
859956   566256     20682     RED RETROSPOT CHILDRENS UMBRELLA         6   
859954   566256     22730           ALARM CLOCK BAKELIKE IVORY         4   
859957   566256     37495       FAIRY CAKE BIRTHDAY CANDLE SET         4   
888248   568518     22728            ALARM CLOCK BAKELIKE PINK         4   
888247   568518     22726           ALARM CLOCK BAKELIKE GREEN         4   
888250   568518      POST                              POSTAGE         1   
888246   568518     22635     CHILDS BREAKFAST SET DOLLY GIRL          2   
888249   568518     22727            ALARM CLOCK BAKELIKE RED          4   
888244   568518     22629                  SPACEBOY LUNCH BOX         12   
888243   568518     22630                 DOLLY GIRL LUNCH BOX        12   
888242   568518     22556       PLASTERS IN TIN CIRCUS PARADE         12   
888241   568518     22555            PLASTERS IN TIN STRONGMAN        12   
888245   568518     22634       CHILDS BREAKFAST SET SPACEBOY          2   
1067363  581587     22629                  SPACEBOY LUNCH BOX         12   
1067368  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
1067367  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
1067366  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
1067365  581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
1067364  581587     23256          CHILDRENS CUTLERY SPACEBOY          4   
1067362  581587     22367      CHILDRENS APRON SPACEBOY DESIGN         8   
1067355  581587     22631             CIRCUS PARADE LUNCH BOX         12   
1067360  581587     22726           ALARM CLOCK BAKELIKE GREEN         4   
1067359  581587     22727            ALARM CLOCK BAKELIKE RED          4   
1067358  581587     22728            ALARM CLOCK BAKELIKE PINK         4   
1067357  581587     22555            PLASTERS IN TIN STRONGMAN        12   
1067356  581587     22556       PLASTERS IN TIN CIRCUS PARADE         12   
1067369  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
1067361  581587     22730           ALARM CLOCK BAKELIKE

In [120]:
retail_data["CohortQuarter"]=retail_data['CohortQuarterID'].map(dict(zip(quarters_map.values(),quarters_map.keys())))
retail_data.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  InvoiceQuarterID  \
45228 2010-01-04 09:24:00  22.50       Q1/2010                 0   
45229 2010-01-04 09:43:00  -4.25       Q1/2010                 0   
45230 2010-01-04 09:53:00  22.50       Q1/2010                 0   
45238 2010-01-04 10:28:00  83.40       Q1/2010                 0   
45237 2010-01-04 10:28:00   8.50       Q1/2010                 0   

       CohortQuarterID CohortQuarter  
45228                0       Q1/2010  
45229                0       Q1/2010  
45230                0       Q1/2010  
45238                0       Q1/2010  
45237                0       Q1/2010

In [121]:
retail_data.tail()

Invoice StockCode                     Description  Quantity  \
1067357  581587     22555       PLASTERS IN TIN STRONGMAN        12   
1067356  581587     22556  PLASTERS IN TIN CIRCUS PARADE         12   
1067369  581587     22138   BAKING SET 9 PIECE RETROSPOT          3   
1067361  581587     22730      ALARM CLOCK BAKELIKE IVORY         4   
1067370  581587      POST                         POSTAGE         1   

                 InvoiceDate  Price  Customer ID Country      InvoiceDate_DT  \
1067357  2011-12-09 12:50:00   1.65        12680  France 2011-12-09 12:50:00   
1067356  2011-12-09 12:50:00   1.65        12680  France 2011-12-09 12:50:00   
1067369  2011-12-09 12:50:00   4.95        12680  France 2011-12-09 12:50:00   
1067361  2011-12-09 12:50:00   3.75        12680  France 2011-12-09 12:50:00   
1067370  2011-12-09 12:50:00  18.00        12680  France 2011-12-09 12:50:00   

         Total InvoiceQuater  InvoiceQuarterID  CohortQuarterID CohortQuarter  
1067357  19.80       Q4/2011                 7                6       Q3/2011  
1067356  19.80       Q4/2011                 7                6       Q3/2011  
1067369  14.85       Q4/2011                 7                6       Q3/2011  
1067361  15.00       Q4/2011                 7                6       Q3/2011  
1067370  18.00       Q4/2011                 7                6       Q3/2011

In [122]:
retail_data['CohortIndex']=retail_data["InvoiceQuarterID"]-retail_data['CohortQuarterID']

retail_data.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  InvoiceQuarterID  \
45228 2010-01-04 09:24:00  22.50       Q1/2010                 0   
45229 2010-01-04 09:43:00  -4.25       Q1/2010                 0   
45230 2010-01-04 09:53:00  22.50       Q1/2010                 0   
45238 2010-01-04 10:28:00  83.40       Q1/2010                 0   
45237 2010-01-04 10:28:00   8.50       Q1/2010                 0   

       CohortQuarterID CohortQuarter  CohortIndex  
45228                0       Q1/2010            0  
45229                0       Q1/2010            0  
45230                0       Q1/2010            0  
45238                0       Q1/2010            0  
45237                0       Q1/2010            0

Retention Cohort Analysis
First, I will perform retention analysis and investigate the number of active users. For that, I will group the data, build a pivot table and calculate retention rate in %. Finally, I will plot the triangle chart with Plotly heatmap.

In [123]:
cohort_retention=retail_data.groupby(['CohortQuarterID','CohortIndex'])["Customer ID"].apply(pd.Series.nunique).reset_index()
cohort_retention.rename(columns={'Customer ID':'Customer_Count'},inplace=True)

cohort_retention.head()

CohortQuarterID  CohortIndex  Customer_Count
0                0            0            1866
1                0            1            1152
2                0            2            1071
3                0            3            1148
4                0            4             854

In [124]:
cohort_retention

CohortQuarterID  CohortIndex  Customer_Count
0                 0            0            1866
1                 0            1            1152
2                 0            2            1071
3                 0            3            1148
4                 0            4             854
5                 0            5             853
6                 0            6             858
7                 0            7             852
8                 1            0             960
9                 1            1             396
10                1            2             447
11                1            3             232
12                1            4             314
13                1            5             288
14                1            6             316
15                2            0             632
16                2            1             303
17                2            2             168
18                2            3             169
19                2            4             207
20                2            5             210
21                3            0             831
22                3            1             177
23                3            2             203
24                3            3             168
25                3            4             297
26                4            0             381
27                4            1             161
28                4            2             130
29                4            3             139
30                5            0             324
31                5            1             143
32                5            2             128
33                6            0             402
34                6            1             213
35                7            0             446

In [125]:
cohort_retention_count=cohort_retention.pivot_table(index="CohortQuarterID", 
                                                      columns='CohortIndex', 
                                                      values='Customer_Count')

cohort_retention_count

CohortIndex           0       1       2       3      4      5      6      7
CohortQuarterID                                                            
0                1866.0  1152.0  1071.0  1148.0  854.0  853.0  858.0  852.0
1                 960.0   396.0   447.0   232.0  314.0  288.0  316.0    NaN
2                 632.0   303.0   168.0   169.0  207.0  210.0    NaN    NaN
3                 831.0   177.0   203.0   168.0  297.0    NaN    NaN    NaN
4                 381.0   161.0   130.0   139.0    NaN    NaN    NaN    NaN
5                 324.0   143.0   128.0     NaN    NaN    NaN    NaN    NaN
6                 402.0   213.0     NaN     NaN    NaN    NaN    NaN    NaN
7                 446.0     NaN     NaN     NaN    NaN    NaN    NaN    NaN

In [126]:
cohort_retention_count["CohortQuarter"]=cohort_retention_count.index.map(dict(zip(quarters_map.values(),quarters_map.keys())))

cohort_retention_count

CohortIndex           0       1       2       3      4      5      6      7  \
CohortQuarterID                                                               
0                1866.0  1152.0  1071.0  1148.0  854.0  853.0  858.0  852.0   
1                 960.0   396.0   447.0   232.0  314.0  288.0  316.0    NaN   
2                 632.0   303.0   168.0   169.0  207.0  210.0    NaN    NaN   
3                 831.0   177.0   203.0   168.0  297.0    NaN    NaN    NaN   
4                 381.0   161.0   130.0   139.0    NaN    NaN    NaN    NaN   
5                 324.0   143.0   128.0     NaN    NaN    NaN    NaN    NaN   
6                 402.0   213.0     NaN     NaN    NaN    NaN    NaN    NaN   
7                 446.0     NaN     NaN     NaN    NaN    NaN    NaN    NaN   

CohortIndex     CohortQuarter  
CohortQuarterID                
0                     Q1/2010  
1                     Q2/2010  
2                     Q3/2010  
3                     Q4/2010  
4                     Q1/2011  
5                     Q2/2011  
6                     Q3/2011  
7                     Q4/2011

In [127]:
cohort_retention_count=cohort_retention_count.set_index('CohortQuarter')
cohort_retention_count

CohortIndex         0       1       2       3      4      5      6      7
CohortQuarter                                                            
Q1/2010        1866.0  1152.0  1071.0  1148.0  854.0  853.0  858.0  852.0
Q2/2010         960.0   396.0   447.0   232.0  314.0  288.0  316.0    NaN
Q3/2010         632.0   303.0   168.0   169.0  207.0  210.0    NaN    NaN
Q4/2010         831.0   177.0   203.0   168.0  297.0    NaN    NaN    NaN
Q1/2011         381.0   161.0   130.0   139.0    NaN    NaN    NaN    NaN
Q2/2011         324.0   143.0   128.0     NaN    NaN    NaN    NaN    NaN
Q3/2011         402.0   213.0     NaN     NaN    NaN    NaN    NaN    NaN
Q4/2011         446.0     NaN     NaN     NaN    NaN    NaN    NaN    NaN

In [128]:
cohort_size=cohort_retention_count.iloc[:,0]
cohort_size

CohortQuarter
Q1/2010    1866.0
Q2/2010     960.0
Q3/2010     632.0
Q4/2010     831.0
Q1/2011     381.0
Q2/2011     324.0
Q3/2011     402.0
Q4/2011     446.0
Name: 0, dtype: float64

In [129]:
retention=cohort_retention_count.divide(cohort_size,axis=0)

retention

CohortIndex      0         1         2         3         4         5  \
CohortQuarter                                                          
Q1/2010        1.0  0.617363  0.573955  0.615220  0.457663  0.457128   
Q2/2010        1.0  0.412500  0.465625  0.241667  0.327083  0.300000   
Q3/2010        1.0  0.479430  0.265823  0.267405  0.327532  0.332278   
Q4/2010        1.0  0.212996  0.244284  0.202166  0.357401       NaN   
Q1/2011        1.0  0.422572  0.341207  0.364829       NaN       NaN   
Q2/2011        1.0  0.441358  0.395062       NaN       NaN       NaN   
Q3/2011        1.0  0.529851       NaN       NaN       NaN       NaN   
Q4/2011        1.0       NaN       NaN       NaN       NaN       NaN   

CohortIndex           6         7  
CohortQuarter                      
Q1/2010        0.459807  0.456592  
Q2/2010        0.329167       NaN  
Q3/2010             NaN       NaN  
Q4/2010             NaN       NaN  
Q1/2011             NaN       NaN  
Q2/2011             NaN       NaN  
Q3/2011             NaN       NaN  
Q4/2011             NaN       NaN

In [130]:
retention=(retention*100).round(2)
retention

CohortIndex        0      1      2      3      4      5      6      7
CohortQuarter                                                        
Q1/2010        100.0  61.74  57.40  61.52  45.77  45.71  45.98  45.66
Q2/2010        100.0  41.25  46.56  24.17  32.71  30.00  32.92    NaN
Q3/2010        100.0  47.94  26.58  26.74  32.75  33.23    NaN    NaN
Q4/2010        100.0  21.30  24.43  20.22  35.74    NaN    NaN    NaN
Q1/2011        100.0  42.26  34.12  36.48    NaN    NaN    NaN    NaN
Q2/2011        100.0  44.14  39.51    NaN    NaN    NaN    NaN    NaN
Q3/2011        100.0  52.99    NaN    NaN    NaN    NaN    NaN    NaN
Q4/2011        100.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN

In [131]:
retention=retention.iloc[::-1]
retention

CohortIndex        0      1      2      3      4      5      6      7
CohortQuarter                                                        
Q4/2011        100.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN
Q3/2011        100.0  52.99    NaN    NaN    NaN    NaN    NaN    NaN
Q2/2011        100.0  44.14  39.51    NaN    NaN    NaN    NaN    NaN
Q1/2011        100.0  42.26  34.12  36.48    NaN    NaN    NaN    NaN
Q4/2010        100.0  21.30  24.43  20.22  35.74    NaN    NaN    NaN
Q3/2010        100.0  47.94  26.58  26.74  32.75  33.23    NaN    NaN
Q2/2010        100.0  41.25  46.56  24.17  32.71  30.00  32.92    NaN
Q1/2010        100.0  61.74  57.40  61.52  45.77  45.71  45.98  45.66

In [132]:
fig = go.Figure(data=go.Heatmap(
                    z=retention.values,
                    y=retention.index,
                    colorscale='Greens',
                    text=retention,
                    texttemplate = "%{text}%",
                    colorbar_title='Retention Rate, %',
                    xgap=3,
                    ygap=3))

#fig.update_xaxes(side="top")

fig.update_layout(title="Cohort Analysis: Retention Rate",
                  xaxis_title="Cohorts",
                  yaxis_title="Quarters",
                  
                  title_x=0.5,
                  title_y=0.99,
                  plot_bgcolor='white')
fig.show()


The results are quite interesting. Have you noticed how retention drops in 2010? An it seems that the store has implemented some measures to improve the situation in 2011, so more customers returned to shop here agein.

Average Quantity Cohort Analysis
One more dimension I am interested is how many products are sold on average, and how this changes with time.

Here, I will use the same method, but group my data using mean() function.

In [134]:
cohort_quantity=retail_data.groupby(["CohortQuarterID","CohortIndex"])["Quantity"].mean().reset_index()

cohort_quantity.rename(columns={'Quantity':'Average Quantity'},inplace=True)

cohort_quantity.head()

CohortQuarterID  CohortIndex  Average Quantity
0                0            0         15.510993
1                0            1         13.124922
2                0            2         14.435120
3                0            3         13.058323
4                0            4         13.955552

In [136]:
average_quantity=cohort_quantity.pivot_table(index="CohortQuarterID", 
                                                      columns='CohortIndex', 
                                                      values='Average Quantity').round(1)

average_quantity

CohortIndex         0     1     2     3     4     5     6     7
CohortQuarterID                                                
0                15.5  13.1  14.4  13.1  14.0  13.7  15.1  13.5
1                10.3  11.2  10.1  14.8  14.2  15.4   9.7   NaN
2                13.5   7.9   9.6  11.8  13.9  10.2   NaN   NaN
3                 7.6   8.4   8.3   9.7   8.7   NaN   NaN   NaN
4                 9.6  13.4  12.4  11.5   NaN   NaN   NaN   NaN
5                 8.8  11.3   9.2   NaN   NaN   NaN   NaN   NaN
6                 9.2   6.2   NaN   NaN   NaN   NaN   NaN   NaN
7                 9.1   NaN   NaN   NaN   NaN   NaN   NaN   NaN

In [137]:
average_quantity["CohorQuarter"]=average_quantity.index.map(dict(zip(quarters_map.values(),quarters_map.keys())))

average_quantity

CohortIndex         0     1     2     3     4     5     6     7 CohorQuarter
CohortQuarterID                                                             
0                15.5  13.1  14.4  13.1  14.0  13.7  15.1  13.5      Q1/2010
1                10.3  11.2  10.1  14.8  14.2  15.4   9.7   NaN      Q2/2010
2                13.5   7.9   9.6  11.8  13.9  10.2   NaN   NaN      Q3/2010
3                 7.6   8.4   8.3   9.7   8.7   NaN   NaN   NaN      Q4/2010
4                 9.6  13.4  12.4  11.5   NaN   NaN   NaN   NaN      Q1/2011
5                 8.8  11.3   9.2   NaN   NaN   NaN   NaN   NaN      Q2/2011
6                 9.2   6.2   NaN   NaN   NaN   NaN   NaN   NaN      Q3/2011
7                 9.1   NaN   NaN   NaN   NaN   NaN   NaN   NaN      Q4/2011

In [139]:
average_quantity=average_quantity.set_index('CohorQuarter')


In [144]:
average_quantity

CohortIndex      0     1     2     3     4     5     6     7
CohorQuarter                                                
Q4/2011        9.1   NaN   NaN   NaN   NaN   NaN   NaN   NaN
Q3/2011        9.2   6.2   NaN   NaN   NaN   NaN   NaN   NaN
Q2/2011        8.8  11.3   9.2   NaN   NaN   NaN   NaN   NaN
Q1/2011        9.6  13.4  12.4  11.5   NaN   NaN   NaN   NaN
Q4/2010        7.6   8.4   8.3   9.7   8.7   NaN   NaN   NaN
Q3/2010       13.5   7.9   9.6  11.8  13.9  10.2   NaN   NaN
Q2/2010       10.3  11.2  10.1  14.8  14.2  15.4   9.7   NaN
Q1/2010       15.5  13.1  14.4  13.1  14.0  13.7  15.1  13.5

In [142]:
average_quantity=average_quantity.iloc[::-1]

In [143]:
fig = go.Figure(data=go.Heatmap(
                    z=average_quantity,
                    y=average_quantity.index,
                    colorscale='Blues',
                    text=average_quantity,
                    texttemplate = "%{text}",
                    colorbar_title='Average Quantity',
                    xgap=3,
                    ygap=3))

fig.update_xaxes(side="top")

fig.update_layout(title="Cohort Analysis: Average Quantity",
                  xaxis_title="Cohorts",
                  yaxis_title="Quarters",
                  
                  title_x=0.5,
                  title_y=0.99,
                  plot_bgcolor='white')
fig.show()


Here as well, a drop in the average sold quantity of products can be noticed in 2010. In 2011 things started to improve but then dropped back; however it may mean that more expensive products are now sold in smaller amounts.

Online Sales on Timeline

Finally, I will built two more charts to show how the sales amount was changing with time. I will break down data by quarters and by months.

In [147]:
retail_quarters_by_sales = retail_data.groupby(['InvoiceQuarterID'], as_index=False)['Total'].agg('sum')
retail_quarters_by_sales['InvoiceQuarter'] = retail_quarters_by_sales['InvoiceQuarterID'].map(dict(zip(quarters_map.values(), 
                                                                                                               quarters_map.keys())))

In [148]:
fig = go.Figure(data=[
    go.Bar(name='Quarterly Sales', 
           x=retail_quarters_by_sales['InvoiceQuarter'],
           y=retail_quarters_by_sales['Total'],
           marker_opacity=1,
           text='Total',
           marker={'color': retail_quarters_by_sales['Total'],
                   'colorscale': 'Portland'}
          )
    ])

fig.update_traces(texttemplate='%{text}: £%{y:.3s}', textposition='outside')
fig.update_layout(barmode='group', showlegend=False)

fig.update_layout(title="Quarterly Sales",
                  title_x = 0.45,
                  xaxis_title="Quarters",
                  yaxis_title="Total Amount, £",
                  plot_bgcolor='white')
fig.show()

We can easily notice the seasonal fluctuations here: people tend to buy a lot before Christmas, and afterwards, a significant drop happens.

In [149]:
retail_monthly = retail_data.copy()
retail_monthly['InvoiceMonth'] = retail_monthly['InvoiceDate_DT'].dt.to_period('M')
retail_monthly_by_sales = retail_monthly.groupby(['InvoiceMonth'], as_index=False)['Total'].agg('sum')

In [150]:
retail_monthly.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  InvoiceQuarterID  \
45228 2010-01-04 09:24:00  22.50       Q1/2010                 0   
45229 2010-01-04 09:43:00  -4.25       Q1/2010                 0   
45230 2010-01-04 09:53:00  22.50       Q1/2010                 0   
45238 2010-01-04 10:28:00  83.40       Q1/2010                 0   
45237 2010-01-04 10:28:00   8.50       Q1/2010                 0   

       CohortQuarterID CohortQuarter  CohortIndex InvoiceMonth  
45228                0       Q1/2010            0      2010-01  
45229                0       Q1/2010            0      2010-01  
45230                0       Q1/2010            0      2010-01  
45238                0       Q1/2010            0      2010-01  
45237                0       Q1/2010            0      2010-01

In [151]:
fig = go.Figure(data=[
    go.Bar(name='Monthly Sales', 
           x=retail_monthly_by_sales['InvoiceMonth'].astype(str),
           y=retail_monthly_by_sales['Total'],
           marker_opacity=1,
           marker={'color': retail_monthly_by_sales['Total'],
                   'colorscale': 'Portland'}
          )
    ])

fig.update_layout(barmode='group', showlegend=False)

fig.update_layout(title="Monthly Sales",
                  title_x = 0.45,
                  xaxis_title="Months",
                  yaxis_title="Total Amount, £",
                  plot_bgcolor='white')
fig.show()

In [152]:
retail_data_month=retail_data.copy()

In [153]:
retail_data_month['InvoiceMonth']=(retail_data['InvoiceDate_DT'].dt.month.astype(str) + '/' + retail_data['InvoiceDate_DT'].dt.year.astype(str))

retail_data_month.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  InvoiceQuarterID  \
45228 2010-01-04 09:24:00  22.50       Q1/2010                 0   
45229 2010-01-04 09:43:00  -4.25       Q1/2010                 0   
45230 2010-01-04 09:53:00  22.50       Q1/2010                 0   
45238 2010-01-04 10:28:00  83.40       Q1/2010                 0   
45237 2010-01-04 10:28:00   8.50       Q1/2010                 0   

       CohortQuarterID CohortQuarter  CohortIndex InvoiceMonth  
45228                0       Q1/2010            0       1/2010  
45229                0       Q1/2010            0       1/2010  
45230                0       Q1/2010            0       1/2010  
45238                0       Q1/2010            0       1/2010  
45237                0       Q1/2010            0       1/2010

In [154]:
retail_data_month["InvoiceMonth"].unique()


array(['1/2010', '2/2010', '3/2010', '4/2010', '5/2010', '6/2010',
       '7/2010', '8/2010', '9/2010', '10/2010', '11/2010', '12/2010',
       '1/2011', '2/2011', '3/2011', '4/2011', '5/2011', '6/2011',
       '7/2011', '8/2011', '9/2011', '10/2011', '11/2011', '12/2011'],
      dtype=object)

In [155]:
month_map=dict(zip(retail_data_month["InvoiceMonth"].unique(),range(len(retail_data_month['InvoiceMonth'].unique()))))
month_map


{'1/2010': 0,
 '2/2010': 1,
 '3/2010': 2,
 '4/2010': 3,
 '5/2010': 4,
 '6/2010': 5,
 '7/2010': 6,
 '8/2010': 7,
 '9/2010': 8,
 '10/2010': 9,
 '11/2010': 10,
 '12/2010': 11,
 '1/2011': 12,
 '2/2011': 13,
 '3/2011': 14,
 '4/2011': 15,
 '5/2011': 16,
 '6/2011': 17,
 '7/2011': 18,
 '8/2011': 19,
 '9/2011': 20,
 '10/2011': 21,
 '11/2011': 22,
 '12/2011': 23}

In [157]:
retail_data_month['InvoiceMonthID']=retail_data_month['InvoiceMonth'].map(month_map)

retail_data_month.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  InvoiceQuarterID  \
45228 2010-01-04 09:24:00  22.50       Q1/2010                 0   
45229 2010-01-04 09:43:00  -4.25       Q1/2010                 0   
45230 2010-01-04 09:53:00  22.50       Q1/2010                 0   
45238 2010-01-04 10:28:00  83.40       Q1/2010                 0   
45237 2010-01-04 10:28:00   8.50       Q1/2010                 0   

       CohortQuarterID CohortQuarter  CohortIndex InvoiceMonth  InvoiceMonthID  
45228                0       Q1/2010            0       1/2010               0  
45229                0       Q1/2010            0       1/2010               0  
45230                0       Q1/2010            0       1/2010               0  
45238                0       Q1/2010            0       1/2010               0  
45237                0       Q1/2010            0       1/2010               0

In [159]:
retail_data_month["cohortMonthID"]=retail_data_month.groupby('Customer ID')['InvoiceMonthID'].transform('min')
retail_data_month.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  InvoiceQuarterID  \
45228 2010-01-04 09:24:00  22.50       Q1/2010                 0   
45229 2010-01-04 09:43:00  -4.25       Q1/2010                 0   
45230 2010-01-04 09:53:00  22.50       Q1/2010                 0   
45238 2010-01-04 10:28:00  83.40       Q1/2010                 0   
45237 2010-01-04 10:28:00   8.50       Q1/2010                 0   

       CohortQuarterID CohortQuarter  CohortIndex InvoiceMonth  \
45228                0       Q1/2010            0       1/2010   
45229                0       Q1/2010            0       1/2010   
45230                0       Q1/2010            0       1/2010   
45238                0       Q1/2010            0       1/2010   
45237                0       Q1/2010            0       1/2010   

       InvoiceMonthID  cohortMonthID  
45228               0              0  
45229               0              0  
45230               0              0  
45238               0              0  
45237               0              0

In [160]:
retail_data_month['cohortMonth']=retail_data_month['cohortMonthID'].map(dict(zip(month_map.values(),month_map.keys())))


retail_data_month.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  InvoiceQuarterID  \
45228 2010-01-04 09:24:00  22.50       Q1/2010                 0   
45229 2010-01-04 09:43:00  -4.25       Q1/2010                 0   
45230 2010-01-04 09:53:00  22.50       Q1/2010                 0   
45238 2010-01-04 10:28:00  83.40       Q1/2010                 0   
45237 2010-01-04 10:28:00   8.50       Q1/2010                 0   

       CohortQuarterID CohortQuarter  CohortIndex InvoiceMonth  \
45228                0       Q1/2010            0       1/2010   
45229                0       Q1/2010            0       1/2010   
45230                0       Q1/2010            0       1/2010   
45238                0       Q1/2010            0       1/2010   
45237                0       Q1/2010            0       1/2010   

       InvoiceMonthID  cohortMonthID cohortMonth  
45228               0              0      1/2010  
45229               0              0      1/2010  
45230               0              0      1/2010  
45238               0              0      1/2010  
45237               0              0      1/2010

In [161]:
retail_data_month.tail()


Invoice StockCode                     Description  Quantity  \
1067357  581587     22555       PLASTERS IN TIN STRONGMAN        12   
1067356  581587     22556  PLASTERS IN TIN CIRCUS PARADE         12   
1067369  581587     22138   BAKING SET 9 PIECE RETROSPOT          3   
1067361  581587     22730      ALARM CLOCK BAKELIKE IVORY         4   
1067370  581587      POST                         POSTAGE         1   

                 InvoiceDate  Price  Customer ID Country      InvoiceDate_DT  \
1067357  2011-12-09 12:50:00   1.65        12680  France 2011-12-09 12:50:00   
1067356  2011-12-09 12:50:00   1.65        12680  France 2011-12-09 12:50:00   
1067369  2011-12-09 12:50:00   4.95        12680  France 2011-12-09 12:50:00   
1067361  2011-12-09 12:50:00   3.75        12680  France 2011-12-09 12:50:00   
1067370  2011-12-09 12:50:00  18.00        12680  France 2011-12-09 12:50:00   

         Total InvoiceQuater  InvoiceQuarterID  CohortQuarterID CohortQuarter  \
1067357  19.80       Q4/2011                 7                6       Q3/2011   
1067356  19.80       Q4/2011                 7                6       Q3/2011   
1067369  14.85       Q4/2011                 7                6       Q3/2011   
1067361  15.00       Q4/2011                 7                6       Q3/2011   
1067370  18.00       Q4/2011                 7                6       Q3/2011   

         CohortIndex InvoiceMonth  InvoiceMonthID  cohortMonthID cohortMonth  
1067357            1      12/2011              23             19      8/2011  
1067356            1      12/2011              23             19      8/2011  
1067369            1      12/2011              23             19      8/2011  
1067361            1      12/2011              23             19      8/2011  
1067370            1      12/2011              23             19      8/2011

In [163]:
retail_data_month["cohortMonthIndex"]=retail_data_month['InvoiceMonthID']-retail_data_month["cohortMonthID"]

retail_data_month.head()

Invoice StockCode                     Description  Quantity  \
45228   493410   TEST001         This is a test product.         5   
45229  C493411     21539         RETRO SPOTS BUTTER DISH        -1   
45230   493412   TEST001         This is a test product.         5   
45238   493414     21527  RETRO SPOT TRADITIONAL TEAPOT         12   
45237   493414    35001G            HAND OPEN SHAPE GOLD         2   

               InvoiceDate  Price  Customer ID         Country  \
45228  2010-01-04 09:24:00   4.50        12346  United Kingdom   
45229  2010-01-04 09:43:00   4.25        14590  United Kingdom   
45230  2010-01-04 09:53:00   4.50        12346  United Kingdom   
45238  2010-01-04 10:28:00   6.95        14590  United Kingdom   
45237  2010-01-04 10:28:00   4.25        14590  United Kingdom   

           InvoiceDate_DT  Total InvoiceQuater  InvoiceQuarterID  \
45228 2010-01-04 09:24:00  22.50       Q1/2010                 0   
45229 2010-01-04 09:43:00  -4.25       Q1/2010                 0   
45230 2010-01-04 09:53:00  22.50       Q1/2010                 0   
45238 2010-01-04 10:28:00  83.40       Q1/2010                 0   
45237 2010-01-04 10:28:00   8.50       Q1/2010                 0   

       CohortQuarterID CohortQuarter  CohortIndex InvoiceMonth  \
45228                0       Q1/2010            0       1/2010   
45229                0       Q1/2010            0       1/2010   
45230                0       Q1/2010            0       1/2010   
45238                0       Q1/2010            0       1/2010   
45237                0       Q1/2010            0       1/2010   

       InvoiceMonthID  cohortMonthID cohortMonth  cohortMonthIndex  
45228               0              0      1/2010                 0  
45229               0              0      1/2010                 0  
45230               0              0      1/2010                 0  
45238               0              0      1/2010                 0  
45237               0              0      1/2010                 0

In [164]:
retail_data_month.tail()

Invoice StockCode                     Description  Quantity  \
1067357  581587     22555       PLASTERS IN TIN STRONGMAN        12   
1067356  581587     22556  PLASTERS IN TIN CIRCUS PARADE         12   
1067369  581587     22138   BAKING SET 9 PIECE RETROSPOT          3   
1067361  581587     22730      ALARM CLOCK BAKELIKE IVORY         4   
1067370  581587      POST                         POSTAGE         1   

                 InvoiceDate  Price  Customer ID Country      InvoiceDate_DT  \
1067357  2011-12-09 12:50:00   1.65        12680  France 2011-12-09 12:50:00   
1067356  2011-12-09 12:50:00   1.65        12680  France 2011-12-09 12:50:00   
1067369  2011-12-09 12:50:00   4.95        12680  France 2011-12-09 12:50:00   
1067361  2011-12-09 12:50:00   3.75        12680  France 2011-12-09 12:50:00   
1067370  2011-12-09 12:50:00  18.00        12680  France 2011-12-09 12:50:00   

         Total InvoiceQuater  InvoiceQuarterID  CohortQuarterID CohortQuarter  \
1067357  19.80       Q4/2011                 7                6       Q3/2011   
1067356  19.80       Q4/2011                 7                6       Q3/2011   
1067369  14.85       Q4/2011                 7                6       Q3/2011   
1067361  15.00       Q4/2011                 7                6       Q3/2011   
1067370  18.00       Q4/2011                 7                6       Q3/2011   

         CohortIndex InvoiceMonth  InvoiceMonthID  cohortMonthID cohortMonth  \
1067357            1      12/2011              23             19      8/2011   
1067356            1      12/2011              23             19      8/2011   
1067369            1      12/2011              23             19      8/2011   
1067361            1      12/2011              23             19      8/2011   
1067370            1      12/2011              23             19      8/2011   

         cohortMonthIndex  
1067357                 4  
1067356                 4  
1067369                 4  
1067361                 4  
1067370                 4

Monthly retention rate


In [165]:
cohort_retention_month = retail_data_month.groupby(['cohortMonthID', 'cohortMonthIndex'])['Customer ID'].apply(pd.Series.nunique).reset_index()
cohort_retention_month.rename(columns={'Customer ID': 'Customer Count'}, inplace=True)

cohort_retention_month.head()


cohortMonthID  cohortMonthIndex  Customer Count
0              0                 0             786
1              0                 1             305
2              0                 2             365
3              0                 3             342
4              0                 4             334

In [166]:
cohort_retention_count_month = cohort_retention_month.pivot_table(index='cohortMonthID', 
                                                      columns='cohortMonthIndex', 
                                                      values='Customer Count')
cohort_retention_count_month['CohortMonth'] = cohort_retention_count_month.index.map(dict(zip(month_map.values(), 
                                                                                    month_map.keys())))
cohort_retention_count_month = cohort_retention_count_month.set_index('CohortMonth')

cohort_retention_count_month.head()

cohortMonthIndex     0      1      2      3      4      5      6      7   \
CohortMonth                                                                
1/2010            786.0  305.0  365.0  342.0  334.0  330.0  320.0  294.0   
2/2010            502.0  168.0  141.0  176.0  165.0  132.0  133.0  176.0   
3/2010            578.0  154.0  165.0  156.0  147.0  131.0  159.0  198.0   
4/2010            361.0   91.0   80.0   68.0   75.0   88.0  108.0  111.0   
5/2010            296.0   61.0   55.0   55.0   55.0   79.0   70.0   45.0   

cohortMonthIndex     8      9   ...     14     15     16     17     18     19  \
CohortMonth                     ...                                             
1/2010            321.0  363.0  ...  254.0  215.0  269.0  245.0  236.0  230.0   
2/2010            166.0  184.0  ...   91.0  124.0  116.0  107.0  102.0  145.0   
3/2010            202.0  104.0  ...  120.0  118.0  114.0  105.0  120.0  131.0   
4/2010             55.0   45.0  ...   59.0   56.0   48.0   59.0   67.0   78.0   
5/2010             20.0   27.0  ...   31.0   41.0   49.0   51.0   53.0   18.0   

cohortMonthIndex     20     21     22     23  
CohortMonth                                   
1/2010            263.0  236.0  310.0  147.0  
2/2010            145.0  127.0   59.0    NaN  
3/2010            154.0   65.0    NaN    NaN  
4/2010             30.0    NaN    NaN    NaN  
5/2010              NaN    NaN    NaN    NaN  

[5 rows x 24 columns]

In [169]:
cohort_retention_count_month.head(15)

cohortMonthIndex     0      1      2      3      4      5      6      7   \
CohortMonth                                                                
1/2010            786.0  305.0  365.0  342.0  334.0  330.0  320.0  294.0   
2/2010            502.0  168.0  141.0  176.0  165.0  132.0  133.0  176.0   
3/2010            578.0  154.0  165.0  156.0  147.0  131.0  159.0  198.0   
4/2010            361.0   91.0   80.0   68.0   75.0   88.0  108.0  111.0   
5/2010            296.0   61.0   55.0   55.0   55.0   79.0   70.0   45.0   
6/2010            303.0   63.0   59.0   63.0   72.0   92.0   41.0   30.0   
7/2010            203.0   44.0   44.0   57.0   62.0   35.0   27.0   29.0   
8/2010            173.0   40.0   54.0   55.0   31.0   24.0   18.0   25.0   
9/2010            256.0   74.0   67.0   35.0   25.0   26.0   36.0   25.0   
10/2010           406.0  121.0   68.0   50.0   39.0   37.0   53.0   54.0   
11/2010           347.0   73.0   32.0   35.0   30.0   30.0   45.0   39.0   
12/2010            78.0    9.0    4.0    7.0    9.0    6.0    4.0    7.0   
1/2011             75.0   17.0   15.0   14.0   15.0   11.0   11.0   10.0   
2/2011            125.0   25.0   22.0   23.0   27.0   21.0   22.0   20.0   
3/2011            181.0   43.0   44.0   40.0   41.0   31.0   39.0   41.0   

cohortMonthIndex     8      9   ...     14     15     16     17     18     19  \
CohortMonth                     ...                                             
1/2010            321.0  363.0  ...  254.0  215.0  269.0  245.0  236.0  230.0   
2/2010            166.0  184.0  ...   91.0  124.0  116.0  107.0  102.0  145.0   
3/2010            202.0  104.0  ...  120.0  118.0  114.0  105.0  120.0  131.0   
4/2010             55.0   45.0  ...   59.0   56.0   48.0   59.0   67.0   78.0   
5/2010             20.0   27.0  ...   31.0   41.0   49.0   51.0   53.0   18.0   
6/2010             24.0   40.0  ...   33.0   40.0   44.0   68.0   22.0    NaN   
7/2010             29.0   24.0  ...   40.0   36.0   51.0   20.0    NaN    NaN   
8/2010             23.0   22.0  ...   32.0   33.0   11.0    NaN    NaN    NaN   
9/2010             31.0   36.0  ...   49.0   14.0    NaN    NaN    NaN    NaN   
10/2010            44.0   36.0  ...   31.0    NaN    NaN    NaN    NaN    NaN   
11/2010            29.0   30.0  ...    NaN    NaN    NaN    NaN    NaN    NaN   
12/2010             5.0    2.0  ...    NaN    NaN    NaN    NaN    NaN    NaN   
1/2011              8.0   17.0  ...    NaN    NaN    NaN    NaN    NaN    NaN   
2/2011             21.0   25.0  ...    NaN    NaN    NaN    NaN    NaN    NaN   
3/2011             46.0   17.0  ...    NaN    NaN    NaN    NaN    NaN    NaN   

cohortMonthIndex     20     21     22     23  
CohortMonth                                   
1/2010            263.0  236.0  310.0  147.0  
2/2010            145.0  127.0   59.0    NaN  
3/2010            154.0   65.0    NaN    NaN  
4/2010             30.0    NaN    NaN    NaN  
5/2010              NaN    NaN    NaN    NaN  
6/2010              NaN    NaN    NaN    NaN  
7/2010              NaN    NaN    NaN    NaN  
8/2010              NaN    NaN    NaN    NaN  
9/2010              NaN    NaN    NaN    NaN  
10/2010             NaN    NaN    NaN    NaN  
11/2010             NaN    NaN    NaN    NaN  
12/2010             NaN    NaN    NaN    NaN  
1/2011              NaN    NaN    NaN    NaN  
2/2011              NaN    NaN    NaN    NaN  
3/2011              NaN    NaN    NaN    NaN  

[15 rows x 24 columns]

In [167]:
cohort_size_month = cohort_retention_count_month.iloc[:, 0]
retention_month = cohort_retention_count_month.divide(cohort_size_month, axis = 0)
retention_month = (retention_month*100).round(2)
retention_month = retention_month.iloc[::-1]

In [168]:
retention_month

cohortMonthIndex     0      1      2      3      4      5      6      7   \
CohortMonth                                                                
12/2011           100.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
11/2011           100.0  17.01    NaN    NaN    NaN    NaN    NaN    NaN   
10/2011           100.0  33.04  17.41    NaN    NaN    NaN    NaN    NaN   
9/2011            100.0  34.55  38.74  15.71    NaN    NaN    NaN    NaN   
8/2011            100.0  33.33  30.56  27.78  18.52    NaN    NaN    NaN   
7/2011            100.0  23.30  30.10  29.13  34.95  15.53    NaN    NaN   
6/2011            100.0  26.85  22.22  27.78  22.22  28.70  10.19    NaN   
5/2011            100.0  27.27  24.55  18.18  21.82  20.91  26.36  10.91   
4/2011            100.0  26.42  20.75  19.81  19.81  24.53  19.81  25.47   
3/2011            100.0  23.76  24.31  22.10  22.65  17.13  21.55  22.65   
2/2011            100.0  20.00  17.60  18.40  21.60  16.80  17.60  16.00   
1/2011            100.0  22.67  20.00  18.67  20.00  14.67  14.67  13.33   
12/2010           100.0  11.54   5.13   8.97  11.54   7.69   5.13   8.97   
11/2010           100.0  21.04   9.22  10.09   8.65   8.65  12.97  11.24   
10/2010           100.0  29.80  16.75  12.32   9.61   9.11  13.05  13.30   
9/2010            100.0  28.91  26.17  13.67   9.77  10.16  14.06   9.77   
8/2010            100.0  23.12  31.21  31.79  17.92  13.87  10.40  14.45   
7/2010            100.0  21.67  21.67  28.08  30.54  17.24  13.30  14.29   
6/2010            100.0  20.79  19.47  20.79  23.76  30.36  13.53   9.90   
5/2010            100.0  20.61  18.58  18.58  18.58  26.69  23.65  15.20   
4/2010            100.0  25.21  22.16  18.84  20.78  24.38  29.92  30.75   
3/2010            100.0  26.64  28.55  26.99  25.43  22.66  27.51  34.26   
2/2010            100.0  33.47  28.09  35.06  32.87  26.29  26.49  35.06   
1/2010            100.0  38.80  46.44  43.51  42.49  41.98  40.71  37.40   

cohortMonthIndex     8      9   ...     14     15     16     17     18     19  \
CohortMonth                     ...                                             
12/2011             NaN    NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN   
11/2011             NaN    NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN   
10/2011             NaN    NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN   
9/2011              NaN    NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN   
8/2011              NaN    NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN   
7/2011              NaN    NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN   
6/2011              NaN    NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN   
5/2011              NaN    NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN   
4/2011            12.26    NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN   
3/2011            25.41   9.39  ...    NaN    NaN    NaN    NaN    NaN    NaN   
2/2011            16.80  20.00  ...    NaN    NaN    NaN    NaN    NaN    NaN   
1/2011            10.67  22.67  ...    NaN    NaN    NaN    NaN    NaN    NaN   
12/2010            6.41   2.56  ...    NaN    NaN    NaN    NaN    NaN    NaN   
11/2010            8.36   8.65  ...    NaN    NaN    NaN    NaN    NaN    NaN   
10/2010           10.84   8.87  ...   7.64    NaN    NaN    NaN    NaN    NaN   
9/2010            12.11  14.06  ...  19.14   5.47    NaN    NaN    NaN    NaN   
8/2010            13.29  12.72  ...  18.50  19.08   6.36    NaN    NaN    NaN   
7/2010            14.29  11.82  ...  19.70  17.73  25.12   9.85    NaN    NaN   
6/2010             7.92  13.20  ...  10.89  13.20  14.52  22.44   7.26    NaN   
5/2010             6.76   9.12  ...  10.47  13.85  16.55  17.23  17.91   6.08   
4/2010            15.24  12.47  ...  16.34  15.51  13.30  16.34  18.56  21.61   
3/2010            34.95  17.99  ...  20.76  20.42  19.72  18.17  20.76  22.66   
2/2010            33.07  36.65  ...  18.13  24.70  23.11  21.31  20.32  28.88  

In [170]:
fig = go.Figure(data=go.Heatmap(
                    z=retention_month.values,
                    y=retention_month.index,
                    colorscale='Greens',
                    text=retention_month,
                    texttemplate = "%{text}%",
                    colorbar_title='Retention Rate, %',
                    xgap=3,
                    ygap=3))

#fig.update_xaxes(side="top")

fig.update_layout(title="Cohort Analysis: Retention Rate",
                  xaxis_title="Cohorts",
                  yaxis_title="monthly",
                  
                  title_x=0.5,
                  title_y=0.99,
                  plot_bgcolor='white')
fig.show()